ml-algorithms/src/main/java/org/opensearch/ml/engine/ingest/S3DataIngestion.java

```diff
@@ -0,0 +1,207 @@
+import software.amazon.awssdk.regions.Region;
+import software.amazon.awssdk.services.s3.S3Client;
+import software.amazon.awssdk.services.s3.model.GetObjectRequest;
+import software.amazon.awssdk.services.s3.model.GetObjectResponse;
+import software.amazon.awssdk.services.s3.model.S3Exception;
+
+@Log4j2
+@Ingester("s3")
+public class S3DataIngestion extends AbstractIngestion {
+    public static final String SOURCE = "source";
+
+    public S3DataIngestion(Client client) {
+        super(client);
+    }
+
+    @Override
+    public double ingest(MLBatchIngestionInput mlBatchIngestionInput) {
+        S3Client s3 = initS3Client(mlBatchIngestionInput);
+
+        List<String> s3Uris = (List<String>) mlBatchIngestionInput.getDataSources().get(SOURCE);
+        if (Objects.isNull(s3Uris) || s3Uris.isEmpty()) {
+            return 100;
+        }
+        boolean isSoleSource = s3Uris.size() == 1;
+        List<Double> successRates = Collections.synchronizedList(new ArrayList<>());
+        for (int sourceIndex = 0; sourceIndex < s3Uris.size(); sourceIndex++) {
+            successRates.add(ingestSingleSource(s3, s3Uris.get(sourceIndex), mlBatchIngestionInput, sourceIndex, isSoleSource));
+        }
+
+        return calculateSuccessRate(successRates);
+    }
+
+    public double ingestSingleSource(
+        S3Client s3,
+        String s3Uri,
+        MLBatchIngestionInput mlBatchIngestionInput,
+        int sourceIndex,
+        boolean isSoleSource
+    ) {
+        String bucketName = getS3BucketName(s3Uri);
+        String keyName = getS3KeyName(s3Uri);
+        GetObjectRequest getObjectRequest = GetObjectRequest.builder().bucket(bucketName).key(keyName).build();
+        double successRate = 0;
+
+        try {
+            ResponseInputStream<GetObjectResponse> s3is = AccessController
+                .doPrivileged((PrivilegedExceptionAction<ResponseInputStream<GetObjectResponse>>) () -> s3.getObject(getObjectRequest));
+            BufferedReader reader = new BufferedReader(new InputStreamReader(s3is, StandardCharsets.UTF_8));
+
+            List<String> linesBuffer = new ArrayList<>();
+            String line;
+            int lineCount = 0;
+            // Atomic counters for tracking success and failure
+            AtomicInteger successfulBatches = new AtomicInteger(0);
+            AtomicInteger failedBatches = new AtomicInteger(0);
+            // List of CompletableFutures to track batch ingestion operations
+            List<CompletableFuture<Void>> futures = new ArrayList<>();
+
+            while ((line = reader.readLine()) != null) {
+                linesBuffer.add(line);
+                lineCount++;
+
+                // Process every 100 lines
+                if (lineCount == 100) {
+                    // Create a CompletableFuture that will be completed by the bulkResponseListener
+                    CompletableFuture<Void> future = new CompletableFuture<>();
+                    batchIngest(
```

Comment:
This part is not blocking, so it's possbile that multiple bulk request running in parallel. Suggest do some load test to check if any risk.

Reply:
Good point. Will need a more load test to verify.